cli/lsp/tsc.rs

```diff
@@ -0,0 +1,1061 @@
+          )))
+        } else if let Some(dependency) = dependencies.get(specifier) {
+          let resolved_import =
+            if let Some(resolved_import) = &dependency.maybe_type {
+              resolved_import.clone()
+            } else if let Some(resolved_import) = &dependency.maybe_code {
+              resolved_import.clone()
+            } else {
+              ResolvedImport::Err("missing dependency".to_string())
+            };
+          if let ResolvedImport::Resolved(resolved_specifier) = resolved_import
+          {
+            let media_type = if let Some(media_type) =
+              sources.get_media_type(&resolved_specifier)
+            {
+              media_type
+            } else {
+              MediaType::from(&resolved_specifier)
+            };
+            resolved.push(Some((
+              resolved_specifier.to_string(),
+              media_type.as_ts_extension(),
+            )));
+          } else {
+            resolved.push(None);
+          }
+        }
+      }
+    }
+  } else if sources.contains(&referrer) {
+    for specifier in &v.specifiers {
+      if let Some((resolved_specifier, media_type)) =
+        sources.resolve_import(specifier, &referrer)
+      {
+        resolved.push(Some((
+          resolved_specifier.to_string(),
+          media_type.as_ts_extension(),
+        )));
+      } else {
+        resolved.push(None);
+      }
+    }
+  } else {
+    return Err(custom_error(
+      "NotFound",
+      "the referring specifier is unexpectedly missing",
+    ));
+  }
+
+  Ok(json!(resolved))
+}
+
+fn respond(state: &mut State, args: Value) -> Result<Value, AnyError> {
+  state.response = Some(serde_json::from_value(args)?);
+  Ok(json!(true))
+}
+
+fn script_names(state: &mut State, _args: Value) -> Result<Value, AnyError> {
+  let script_names: Vec<&ModuleSpecifier> =
+    state.server_state.doc_data.keys().collect();
+  Ok(json!(script_names))
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct ScriptVersionArgs {
+  specifier: String,
+}
+
+fn script_version(state: &mut State, args: Value) -> Result<Value, AnyError> {
+  let v: ScriptVersionArgs = serde_json::from_value(args)?;
+  let specifier = ModuleSpecifier::resolve_url(&v.specifier)?;
+  let maybe_doc_data = state.server_state.doc_data.get(&specifier);
+  if let Some(doc_data) = maybe_doc_data {
+    if let Some(version) = doc_data.version {
+      return Ok(json!(version.to_string()));
+    }
+  } else {
+    let mut sources = state.server_state.sources.write().unwrap();
+    if let Some(version) = sources.get_script_version(&specifier) {
+      return Ok(json!(version));
+    }
+  }
+
+  Ok(json!(None::<String>))
+}
+
+/// Create and setup a JsRuntime based on a snapshot. It is expected that the
+/// supplied snapshot is an isolate that contains the TypeScript language
+/// server.
+pub fn start(snapshot: Snapshot, debug: bool) -> Result<JsRuntime, AnyError> {
```

Comment:
So it looks like we'll have 2 different implementations of TSC, do they share any code?

Reply:
Also it seems to me `snapshot` argument doesn't make much sense, are we ever gonna start this `JsRuntime` with other type of snapshot?

Reply:
Answering myself to the first question, the code is shared in `99_main_compiler.js` it's just a different set of ops that's used.

Reply:
cleaned up